### PR TITLE
Remove Experience Config Delivery Mechanism [#3322]

### LINKED ISSUE #3322: [Backend] Update PrivacyExperienceConfig fields to support new requirements
### Is your feature request related to a specific problem?

[See design doc here](https://ethyca.atlassian.net/wiki/spaces/EN/pages/2697166893/Handling+Error+Conditions+for+Experience+Configuration+May+2023)

- Removing concept of delivery mechanism, replacing with banner_enabled field
- Removing banner_title and banner_description in favor of component_title and component_description
- Rename link_label to open_modal_link
- Add privacy_policy_link
- Add privacy_policy_url
- Remove validation on how notices must be delivered in a banner

Related to https://github.com/ethyca/fides/issues/3320

### Describe the solution you'd like

_A description of what you want to happen._

### Describe alternatives you've considered, if any

_A description of any alternative solutions or features you've considered._

### Additional context

_Add any other context or screenshots about the feature request here._

### PR DESCRIPTION
**Pull request by pattisdr**
Closes #3322 

❗ Contains migration; check downrev before merging
❗ Blocked by frontend ticket https://github.com/ethyca/fides/issues/3320

### Vocabulary

_Privacy Notice:_   The entity that a user accepts or rejects in terms of consent.  A notice encapsulates a name, consent mechanism, data uses, and potentially many regions.  It is not linked to Privacy Experiences via FK.  Privacy Experiences are potentially created behind the scenes when Privacy Notices are created/updated to make sure we always have experiences to present the given notices.

_Privacy Experience:_  Governs how privacy notices are presented to the end user for a given region. Represents the Experience for a single region and component (overlay or privacy center).  Has a FK to Experience Config.  At runtime, determines what Privacy Notices are relevant for a particular experience and returns.  Components like the privacy center, banners, and overlays should call the Privacy Experience API.

_Experience Config:_  Holds the copy for an experience, such as the overlay's title or button labels.  This ExperienceConfig can be shared between multiple Experiences.  The Admin UI queries the Experience Configs so the customer can set these up.  Experiences are updated behind the scenes when Experience Config updates are made.

### Description Of Changes

Building on the [previous data model changes](https://github.com/ethyca/fides/pull/3292) to [keep Privacy Notices in sync with Privacy Experiences](https://ethyca.atlassian.net/wiki/spaces/EN/pages/2683830495/2023-05-06+Backend+Spike+Keeping+Notices+and+Experiences+in+Sync), this PR simplifies the constraints.

Follows what @NevilleS [proposes here](https://ethyca.atlassian.net/wiki/spaces/EN/pages/2697166893/Handling+Error+Conditions+for+Experience+Configuration+May+2023), primarily by removing `ExperienceConfig.delivery_mechanism` and all associated validation.  Previously, PrivacyNotices that had `opt_in` or `notice_only` consent mechanisms had to be presented in an overlay that was delivered by a **banner** (field stored on the PrivacyExperienceConfig).  There was a lot of behind the scenes activity to ensure that we always kept our PrivacyExperiences in the correct state, when both PrivacyNotices and PrivacyExperienceConfigs were updated. We remove this here and instead surface a property calculated at runtime for the frontend to determine if they need to show a banner.

Similarly makes the related changes to update the ExperienceConfig with the [new fields per updated requirements](https://www.figma.com/file/dmEwdK3xZwjKfGVQ9t66xe/Fides-v.2-Master-Working-Files?type=design&node-id=5115-139765&t=GsLD1WwrrU5ZmeGS-0)

### Code Changes

* [ ] In the GET Privacy Experience List endpoint, surface a new field `PrivacyExperience.show_banner` which is calculated at runtime as to whether or not the given experience needs to be triggered via a banner
* [ ] Remove checking associated privacy notice consent mechanisms when updating or creating ExperienceConfigs to make sure there's not a delivery mechanism mismatch.  Reduces the links between the two.
    * [ ] Remove automatically updating Experiences behind the scenes when PrivacyNotices were adjusted in a way that forced the delivery_mechanism to change
* [ ] Add/rename many fields on the `ExperienceConfig` table to reflect the new mockups.
* [ ] Allow ExperienceConfigs to be updated _without_ passing in a regions key, to support the first release, where the copy can be updated, but the regions can't.  If no regions key supplied, Experiences that are linked to the ExperienceConfig won't be touched.
    * [ ] Add new validation that prevents `ExperienceConfig.component` from being updated once it's created to avoid a bunch of extra backend messiness.
* [ ] Bug fix that changes when we validate `PrivacyNotice` component and `displayed_in_*` fields - there were edge cases where we would be validating too early.  We really need the existing PrivacyNotice object to already be loaded before we validate.

### Steps to Confirm

* [ ] POST `{{host}}/privacy-notice` to create some notices
* [ ] GET  `{{host}}/privacy-experience/` to verify Privacy Experiences still created behind the scenes when the Privacy Experiences were created.  For now `experience_config` should be null.  Verify that `show_banner` is True if there is a notice under `privacy_notices` that has a `consent_mechanism` of `opt_in` or `notice_only`.  `privacy_center` notices however, always have `show_banner` = False
* [ ] POST `{{host}}/experience-config` where your request body has `component`=overlay, and use a region that matches one of your regions in experiences
* [ ] `GET {{host}}/privacy-experience/` and verify that `experience_config` is no longer null for the matching overlay type and region.
* [ ] Do a PATCH {{host}}/experience-config/{{experience-config-id}} to an experience config that was associated with an experience whose "show_banner" was True.  In the request body, set `banner_enabled`: `always_disabled`.  Verify that this should flip `show_banner` to False, for Experiences it is attached to (`GET {{host}}/privacy-experience/` to verify 

### Pre-Merge Checklist

* [ ] All CI Pipelines Succeeded
* Documentation:
  * [ ] documentation complete, [PR opened in fidesdocs](https://github.com/ethyca/fidesdocs/pulls)
  * [ ] documentation [issue created in fidesdocs](https://github.com/ethyca/fidesdocs/issues/new/choose)
* [ ] Issue Requirements are Met
* [ ] Relevant Follow-Up Issues Created
* [x] Update `CHANGELOG.md`
* [x] For API changes, the [Postman collection](https://github.com/ethyca/fides/blob/main/docs/fides/docs/development/postman/Fides.postman_collection.json) has been updated

### CODE REVIEW

File: src/fides/api/models/privacy_experience.py
Comment (pattisdr) [inline src/fides/api/models/privacy_experience.py:38]: A primary change of this PR; `delivery_mechanism` has been replaced in favor of `banner_enabled` which is similar in concept, however this construct further reduces the connections between a Notice and an Experience.    Previously, we did a lot of behind the scenes validation to make sure an experience's delivery_mechanism always matched its notices.  This new field doesn't have that validation, we don't check this at all, so it's possible `ExperienceConfig.banner_enabled` = `always_disabled` regardless of how the underlying notices should be displayed.

File: src/fides/api/models/privacy_experience.py
Comment (pattisdr) [inline src/fides/api/models/privacy_experience.py:43]: Changing to `acknowledge_button_label`

File: src/fides/api/models/privacy_experience.py
Comment (pattisdr) [inline src/fides/api/models/privacy_experience.py:45]: Seems like there was a simplification in the UI that we can just define a title and a description, and they can both be used for a banner title and an overlay title, so I'm removing the duplicate set of fields.

File: src/fides/api/models/privacy_experience.py
Comment (pattisdr) [inline src/fides/api/models/privacy_experience.py:48]: Simplifying to title/description instead.

File: src/fides/api/models/privacy_experience.py
Comment (pattisdr) [inline src/fides/api/models/privacy_experience.py:492]: Removing some of this validation and behind-the-scenes efforts now that this requirement has been removed, as I'm updating PrivacyExperiences to make sure we always have the appropriate rows to back PrivacyNotices.

File: src/fides/api/schemas/privacy_experience.py
Comment (pattisdr) [inline src/fides/api/schemas/privacy_experience.py:26]: I've tried to mark here how the fields are shared between the Overlay Experience Config and the Privacy Center Experience Config.

File: src/fides/api/schemas/privacy_experience.py
Comment (pattisdr) [inline src/fides/api/schemas/privacy_experience.py:90]: Updating the logic here with my understanding of what fields are required when saving an overlay.  Currently overlays and privacy center experience configs are stored in the same table, as I feel there are a lot of overlapping fields. The Privacy Center needs a subset of the Overlay fields.

File: src/fides/api/schemas/privacy_experience.py
Comment (pattisdr) [inline src/fides/api/schemas/privacy_experience.py:119]: New piece of validation here but I think it makes sense to not allow an ExperienceConfig's component to be updated once it's created.  We either do this, or add more behind the scenes logic to unlink more experiences as the ExperienceConfig.component is updated.

File: src/fides/api/schemas/privacy_experience.py
Comment (pattisdr) [inline src/fides/api/schemas/privacy_experience.py:114]: As mentioned elsewhere, regions are no longer required in the request to update the ExperienceConfig. If you just want to update the copy and not the regions, don't include the regions key in the request at all.

File: src/fides/api/schemas/privacy_notice.py
Comment (pattisdr) [inline src/fides/api/schemas/privacy_notice.py:89]: The diff here looks strange but I mostly moved the `validate_consent_mechanisms_and_display` under PrivacyNoticeCreation instead of the base `PrivacyNotice`

File: src/fides/api/api/v1/endpoints/privacy_experience_config_endpoints.py
Comment (allisonking) [reply src/fides/api/api/v1/endpoints/privacy_experience_config_endpoints.py:214]: ah cool, that means I can get rid of this bit in the FE 😄 https://github.com/ethyca/fides/blob/main/clients/admin-ui/src/features/privacy-experience/privacy-experience.slice.ts#L45-L47

File: src/fides/api/models/privacy_experience.py
Comment (allisonking) [reply src/fides/api/models/privacy_experience.py:43]: nice, that lets us not have to suffer from the question `acknowledgment` or `acknowledgement` 😄

File: src/fides/api/schemas/privacy_experience.py
Comment (allisonking) [reply src/fides/api/schemas/privacy_experience.py:26]: this was very helpful, thank you!

**Review by allisonking — COMMENTED** [review]
nice job! was straightforward to integrate with 👍

File: src/fides/api/schemas/privacy_experience.py
Comment (adamsachs) [inline src/fides/api/schemas/privacy_experience.py:113]: this now seems to be identical to what you've got in the `ExperienceConfigCreate` model, including the validator method. is there still a reason to have it split out here in the subclasses and not the base class? i see you've got a field `default` and `description` specified here that may not apply to the "create". still feels like there's some room to DRY this up a bit, perhaps?

File: src/fides/api/models/privacy_experience.py
Comment (adamsachs) [inline src/fides/api/models/privacy_experience.py:221]: somewhat related to my comment below about considering whether we still need this db table/model - it just struck me as a bit odd that this property is on the db model. i can see why you've done it, and i'm not saying we should change it now...just some food for thought.

File: src/fides/api/models/privacy_experience.py
Comment (adamsachs) [inline src/fides/api/models/privacy_experience.py:537]: yet another comment that i'm not suggesting we address in this PR 😄

as i wrap my head around all of this, part of me is starting to feel like the `PrivacyExperience` db model may be a bit superfluous? again, it makes total sense to have this as an API-level model - i like how simple it is as an API consumer to get this simple representation of experiences, all of which have been calculated by some relatively complex logic/maintenance on the BE. but do we get a lot of benefit from keeping this maintained in a db table?

i get the need to audit things and maintain a detailed history, so that could be one motivation -- but i still feel like the audit history could be pieced together by just the `PrivacyNoticeHistory` and `PrivacyExperienceConfigHistory` tables? we'd maybe need to provide some "utility" endpoints to make it easier on the consumer to piece that all together....

i guess what i get worried about here is the maintaining duplicate state across the `PrivacyExperience` and `PrivacyExperienceConfig` _and_ their associated history tables. i went down this line of thinking because it struck me as a bit funny that we're keeping a list of regions on the `PrivacyExperienceConfig` while _also_ maintaining links from `PrivacyExperience` records, whose main purpose is just to record the experience type for a given region (admittedly, that's a bit of an oversimplification). the method you've got here is pretty easy to follow, but we could potentially remove the need for it (and the `upsert_privacy_experiences_after_notice_update` function) and just rely on runtime lookups across the `PrivacyExperienceConfig` and `PrivacyNotice` tables to populate a `PrivacyExperience` API object that looks identical to what you have now - without maintaining a DB table (and associated history table).

i haven't totally thought this through! i can definitely see many good reasons to keep this in place. but it's a line of thinking i wanted to at least entertain as i am making my way through this. let me know what you think!

**Review by adamsachs — APPROVED** [review]
-- note: my review is scoped to the BE changes, i didn't take a look at any FE code changes (though i will try to do some manual testing!) -- 

this looks great @pattisdr! as far as i can tell, this iteration greatly simplifies things. after getting a bit of context to get back up to speed on the use case, i was able to follow just about everything that's going on here - which is a testament to how well you've structured the code, given the complex requirements.

i've really got only a few minor, nonblocking comments that may be worth acting upon in the scope of this PR - so i will preemptively approve this. i do have a few more meaty questions on the data structures/db models that i do not expect to be addressed in this iteration, but that i'd like to get your thinking on.

i haven't done any manual testing, but i'll look to spin that up a bit locally, if only to immerse myself a bit more in all the great work you and the rest of the team have put in place 😄 i'm not letting that hold up an approval here, since i'm assuming you and others have done enough of that manual testing to feel comfortable merging. if that's not the case, please let me know and i can be sure to do that testing before we decide to merge!

**Comment by pattisdr** [comment]
thanks very much for taking the time to review @adamsachs ⭐

File: src/fides/api/models/privacy_experience.py
Comment (pattisdr) [reply src/fides/api/models/privacy_experience.py:43]: hahaha

File: src/fides/api/models/privacy_experience.py
Comment (pattisdr) [reply src/fides/api/models/privacy_experience.py:41]: Thanks for your comments I think @NevilleS also would have stored in a JSON blob, so maybe it's two against one 😄 

My reasoning for storing fields separately:

- Originally experience fields had more "rules" on them, much more upfront validation needed
- Model has evolved a lot, was much more tightly linked to Notices previously. Right now it is more or less contains the "copy", but I'm not convinced that will hold for future iterations as we continue to debate this feature and roll out more functionality
- Easier to validate
- Unknown future filtering and search requirements.  Not a big deal to search json blobs for filtering in the admin ui but what about Consent Reporting search? If we end up needing to search privacypreferencehistory or currentprivacypreference for related experience information I'd rather have these columns separate so I can index.
- I likely would want to keep the jsonb fields clean regardless, so might end up writing migrations to update jsonb still.
- Easier to store with more structure initially and collapse into a json field later than the reverse

Anyway, if it settles out that the ExperienceConfig is truly just copy and we don't have performance concerns with search, that might be a reason to make this a jsonb field instead.

Comment (adamsachs) [reply src/fides/api/models/privacy_experience.py:41]: that's very good reasoning! thanks for spelling it out. totally good with leaving it how it is for now based on those considerations.

File: src/fides/api/models/privacy_experience.py
Comment (pattisdr) [reply src/fides/api/models/privacy_experience.py:164]: Good callout, I've been copying and pasting your original method all over the place with only minor modifications. I'll dry that up now.

File: src/fides/api/models/privacy_experience.py
Comment (pattisdr) [reply src/fides/api/models/privacy_experience.py:245]: This is a current expected state!  Another [ticket this sprint](https://github.com/ethyca/fides/issues/3305) will improve this by adding two default `ExperienceConfigs` to the db on startup, one overlay, and one privacy center, to which I can automatically link a `PrivacyExperience` on the user's behalf where needed.

I do think we'll need further product discussion around "what I should do automatically" versus where we should surface required actions needed to the customer:

Let me give you some examples:
- A customer creates a CA Privacy Notice that should be displayed in an "overlay", I create a CA overlay Privacy Experience in the background.  Right now there's no copy to link this to automatically.
- Another example from the other end, say I have an overlay Experience Config with VA on it.  Say the user removes VA. Now the VA Privacy Experience is not linked to any copy.

File: src/fides/api/models/privacy_experience.py
Comment (pattisdr) [reply src/fides/api/models/privacy_experience.py:263]: What do you know, a `Set` does work with `in_`! Change made.

File: src/fides/api/models/privacy_experience.py
Comment (pattisdr) [reply src/fides/api/models/privacy_experience.py:235]: Thank you! Readability is what I was going for!

File: src/fides/api/ctl/migrations/versions/deb97d9393f3_remove_delivery_mechanism.py
Comment (pattisdr) [reply src/fides/api/ctl/migrations/versions/deb97d9393f3_remove_delivery_mechanism.py:270]: Unless you want to migrate your JSON blob too and keep things clean!

File: src/fides/api/schemas/privacy_experience.py
Comment (pattisdr) [reply src/fides/api/schemas/privacy_experience.py:113]: yes room for cleanup there! I'd originally treated regions differently for create and update in an earlier draft of this PR and later consolidated the behavior to reduce confusion.

File: src/fides/api/models/privacy_experience.py
Comment (pattisdr) [reply src/fides/api/models/privacy_experience.py:537]: To clarify one item up front, regions are _not_ duplicated across the `PrivacyExperienceConfig` and the `PrivacyExperience` models.  PrivacyExperience has a FK to PrivacyExperienceConfig but there is no corresponding list of regions on `PrivacyExperienceConfig`.   There is a `PrivacyExperienceConfig.regions` property but it's calculated from the PrivacyExperience FK relationship.

File: src/fides/api/models/privacy_experience.py
Comment (pattisdr) [reply src/fides/api/models/privacy_experience.py:221]: I do like how the Experience has the all the details for a single region.  `show_banner` really is a property of a single region.  It's looking at notices for that region, and then its ExperienceConfig via the FK to see if this specific region should surface a banner!

File: src/fides/api/ctl/migrations/versions/deb97d9393f3_remove_delivery_mechanism.py
Comment (adamsachs) [reply src/fides/api/ctl/migrations/versions/deb97d9393f3_remove_delivery_mechanism.py:270]: good point, i was getting ahead of myself 😬

File: src/fides/api/schemas/privacy_experience.py
Comment (adamsachs) [reply src/fides/api/schemas/privacy_experience.py]: looks good, thanks for addressing!

File: src/fides/api/models/privacy_experience.py
Comment (adamsachs) [reply src/fides/api/models/privacy_experience.py:510]: > To clarify one item up front, regions are _not_ duplicated across the `PrivacyExperienceConfig` and the `PrivacyExperience` models. PrivacyExperience has a FK to PrivacyExperienceConfig but there is no corresponding list of regions on `PrivacyExperienceConfig`. There is a `PrivacyExperienceConfig.regions` property but it's calculated from the PrivacyExperience FK relationship.

aha, i think i missed this detail since that particular model attribute wasn't being changed in this PR - but it is an important detail that does settle the main source of my qualms, so thank you for recognizing my confusion and pointing this out!

> One of the main purposes of the Privacy Experience table at this point, in my mind, is to enforce that we only have one experience type per region at all times down to the database level.

this makes sense, and especially with the context you've provided on the logical nightmare that existed before (some of which is outside of this PR/discussion). i do think that at least some of the problems you were running into before are probably no longer as pertinent given the delivery mechanism removal you've got in this PR -- but i can certainly better see the motivation to have this sort of join/constraint table, given that changes from "both sides" (i.e. the `PrivacyNotice` and the `PrivacyExperienceConfig`) could prompt an update to what experience is shown for a given reason.

i do think that moving forward we need to be careful to _not_ duplicate data across the tables (i.e. `PrivacyNotice`, `PrivacyExperience` and `PrivacyExperienceConfig`). you likely have a much clearer picture on all of this than i do (still getting up to speed!), but given how interrelated these concepts are, i could imagine there being some confusion about what attributes belong where, as the data model evolves - so we'll need to take care when making updates (as you've done here!).

all that being said, i've been convinced that splitting the tables is a sensible and well-motivated choice. thank you for continuing to provide the context behind all of this!

**Review by adamsachs — APPROVED** [review]
looking great, really awesome work here @pattisdr. 

thank you for the time and care in addressing all my comments - both the adjustments as well as the reasoning and context behind some of the more _philosophical_ questions! 😄

File: src/fides/api/models/privacy_experience.py
Comment (pattisdr) [reply src/fides/api/models/privacy_experience.py]: Naming made sense to me!

**Comment by pattisdr** [comment]
Thanks again for getting up to speed to get this reviewed Adam! Especially appreciate the comments about keeping an eye on what fields are stored in which model and will keep a watch on this one.

**Pull request merged**
